app/packages/looker/src/worker/threed-label-processor.ts

```diff
@@ -0,0 +1,206 @@
+import { DETECTIONS } from "@fiftyone/utilities";
+import { DetectionLabel } from "../overlays/detection";
+import { Sample } from "../state";
+import { mapId } from "./shared";
+
+type DetectionsLabel = {
+  _cls: "Detections";
+  detections: DetectionLabel[];
+};
+
+type ThreeDLabel = DetectionsLabel | DetectionLabel;
+
+type LabelId = string;
+
+const COLLECTION_TYPES = new Set(["Detections"]);
+
+const scalingFactorCache: Record<
+  LabelId,
+  {
+    scalingFactor?: { xScale: number; yScale: number };
+  }
+> = {};
+
+/**
+ * Get scaling parameters from pointcloud bound range.
+ *
+ * Cache results of this function because it is called for every label in a sample.
+ */
+const getScalingFactorForLabel = (
+  labelId: LabelId,
+  width: number,
+  height: number,
+  xmin: number,
+  xmax: number,
+  ymin: number,
+  ymax: number
+) => {
+  if (scalingFactorCache[labelId]?.scalingFactor) {
+    return scalingFactorCache[labelId].scalingFactor;
+  }
+
+  if (!scalingFactorCache[labelId]) {
+    scalingFactorCache[labelId] = {};
+  }
+
+  scalingFactorCache[labelId].scalingFactor = {
+    xScale: width / (xmax - xmin),
+    yScale: height / (ymax - ymin),
+  };
+
+  return scalingFactorCache[labelId].scalingFactor;
+};
+
+// cache between sample id and inferred projection params
+const inferredParamsCache: Record<
+  Sample["id"],
+  Sample["orthographic_projection_metadata"]
+> = {};
+
+/**
+ * Use label attributes to infer width, height, and bounds.
+ */
+const getInferredParamsForUndefinedProjection = (sample: Readonly<Sample>) => {
+  if (inferredParamsCache[sample.id]) {
+    return inferredParamsCache[sample.id];
+  }
+
+  let minX = Infinity;
+  let maxX = -Infinity;
+  let minY = Infinity;
+  let maxY = -Infinity;
+
+  for (const [_, label] of Object.entries(sample)) {
+    if (typeof label !== "object" || !("_cls" in label)) {
+      continue; // skip non-labels like "filepath", "id"
+    }
+
+    if (label._cls === DETECTIONS) {
+      for (const detection of label.detections as DetectionLabel[]) {
+        const [x, y] = detection.location;
+        const [lx, ly] = detection.dimensions;
+
+        minX = Math.min(minX, x - lx / 2);
+        maxX = Math.max(maxX, x + lx / 2);
+        minY = Math.min(minY, y - ly / 2);
+        maxY = Math.max(maxY, y + ly / 2);
+      }
+    } else if (label._cls === "Detection") {
+      const [x, y] = label.location as DetectionLabel["location"];
+      const [lx, ly] = label.dimensions as DetectionLabel["dimensions"];
+
+      minX = Math.min(minX, x - lx / 2);
+      maxX = Math.max(maxX, x + lx / 2);
+      minY = Math.min(minY, y - ly / 2);
+      maxY = Math.max(maxY, y + ly / 2);
+    }
+  }
+
+  const padding = 100;
```

Comment:
nit: move to constant? I have little context but will these numbers be configurable by users in the future?